### ISSUE TITLE
TryApplyChanges fails (returns false) when refactoring a file that is open, but has not been activated.

### ISSUE THREAD

**Issue by kfertitta** [dotnet/roslyn#25425]
**Version Used**: 
2.7.0

**Steps to Reproduce**:

1. Create a VS extension that does some kind of edit to a syntax tree in a file in a VisualStudioWorkspace.
2. Add a command to the extension to trigger the edit operation.
3. Run the extension in the debugger and set a breakpoint on your call to `TryApplyChanges`.
4. Create a C# project with at least two files in it -- say, `File1.cs` and `File2.cs`.
5. Open both files in the editor, select the `File1.cs` tab in the document well so that it's the active document.
6. Close the solution (don't just kill the debugger) and exit the debugged VS instance (don't just kill the debugger).
7. Restart debugging and open the C# project.  `File1.cs` and `File2.cs` should be open and `File1.cs` should be activated.  DON'T click on `File2.cs`, yet.
8. Execute your command that operates on `File2.cs`.
9. Note that `TryApplyChanges` returns false.
10. Now click on `File2.cs` in the document well to activate it.
11. Re-execute the command that operates on `File2.cs`.
12. Note that `TryApplyChanges` succeeds.

**Expected Behavior**:
`TryApplyChanges` should always work unless the `Solution` you supply is somehow out-of-date.

**Actual Behavior**:
The actual behavior is that any file that has been opened from the `.suo` but not yet activated cannot be updated via the syntax editing facilities, unless the edit operations are done *twice*.  Some investigation shows that the first (failing) call to `TryApplyChanges` triggers some sort of solution update because we see a `WorkspaceChanged` event fire, and that event reports a change to the non-activated document that's causing the problem.  But, that event seems to be coming too late and seems to be coming from an async completion somewhere.  Suspect some sort of RDT/async interaction issue is at play here.

I have a repro VSPackage if desired, but it happens every time.

Discussed this with @jasonmalinowski and he indicated this should be reported here.

**Comment by jasonmalinowski** [comment]
FYI @jinujoseph this is user visible -- some lightbulbs or other refactorings might silently fail to apply if you had any tabs that were "lazy loaded" and hadn't been initialized. I suspect this bug isn't new, but something to watch for if we are seeing user feedback.

**Comment by kfertitta** [comment]
Hi @jinujoseph  and @jasonmalinowski, any suggestions for a workaround?  I tried programmatically calling `IVsWindowFrame.Show` on any doc windows I found from the RDT that reported not being initialized, but that doesn't seem to work (at least not consistently).  And, it's very hacky at best -- definitely couldn't ship that way.  I also tried opening/closing each unitialized doc in the invisible editor, but that didn't help.  I haven't tried just programmatically closing and re-opening because that would be even more hacky and not really a viable workaround even if made to work.

**Comment by jasonmalinowski** [comment]
@kfertitta What happens if you use IVsRunningDocumentTable to enumerate all the documents, and call [GetDocumentData](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivsrunningdocumenttable4.getdocumentdata?view=visualstudiosdk-2017) on each cookie? You don't have to do anything from there I think, just call it.

**Comment by kfertitta** [comment]
@jasonmalinowski Thanks for that suggestion.  I did give it a try, but still got the same failure.  It's a shame cuz that would have been a very simple and transparent workaround.

**Comment by jasonmalinowski** [comment]
This bug is happening because the call to ensure files are editable with source control is implicitly opening these files. This causes the Workspace version to move forward so we fail to apply the edit.

I can think of a few fixes here:

1. Do the version check _before_ we go check out the files. Right now today TryApplyChanges in the VisualStudioWorkspace is implemented by overriding and the calling to the base, and there's not a hook I see to do something in the middle. We'll have to add one.
2. Change the logic that updates the version number to not bump it if the only thing that happened is opening a file and the underlying contents were the same. This is tricky and I'm not sure how I feel about that.

This doesn't happen in Visual Studio 2019 because we're deferring opening files from the workspace on an RDT event, so the file open will get processed after TryApplyChanges completes. If we fix that deferral (which we have other reasons to do), then this bug will pop up again.

**Comment by sapsari** [comment]
I have a VS extension that applies multiple CSharpSyntaxRewriters sequentially, updating workspace with TryApplyChanges after each rewriter. It works fine on my PC but one user reported that the extension partially works. When digging his log files I noticed TryApplyChanges fails for some rewriters, in an inconsistent way, like between %40 of time. After seeing this issue I asked the user to close to-be refactored files before using extension and he said problems are gone.

User's VS version is 15.9.12 and extension's Roslyn packages' version is 2.10.0 . I tried to simulate user's and this post's issue with the same VS version but failed.
